dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -0,0 +1,261 @@
+    decoder_info.ulMaxHeight = Height();
+    decoder_info.ulMaxWidth = Width();
+    decoder_info.ulTargetHeight = Height();
+    decoder_info.ulTargetWidth = Width();
+    decoder_info.ulNumDecodeSurfaces = num_decode_surfaces_;
+    decoder_info.ulNumOutputSurfaces = 2;
+
+    CUDA_CALL(cuvidCreateDecoder(&nvdecode_state_->decoder_, &decoder_info));
+
+    // Create nv parser
+    CUVIDPARSERPARAMS parser_info;
+    memset(&parser_info, 0, sizeof(CUVIDPARSERPARAMS));
+    parser_info.CodecType = cudaVideoCodec_H264;
+    parser_info.ulMaxNumDecodeSurfaces = num_decode_surfaces_;
+    parser_info.ulMaxDisplayDelay = 0;
+    parser_info.pUserData = this;
+    parser_info.pfnSequenceCallback = detail::process_video_sequence;
+    parser_info.pfnDecodePicture = detail::process_picture_decode;
+    parser_info.pfnDisplayPicture = nullptr;
+
+    CUDA_CALL(cuvidCreateVideoParser(&nvdecode_state_->parser_, &parser_info));
+
+    // Init internal frame buffer
+    for (size_t i = 0; i < frame_buffer_.size(); ++i) {
+      frame_buffer_[i].frame_.resize(FrameSize());
+      frame_buffer_[i].pts_ = -1;
+    }
+}
+
+void FramesDecoderGpu::SeekFrame(int frame_id) {
+  SendLastPacket(true);
+  FramesDecoder::SeekFrame(frame_id);
+}
+
+bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
+  // No more frames in the file
+  if (current_frame_ == -1) {
+    return false;
+  }
+
+  // Check if requested frame was buffered earlier
+  for (auto &frame : frame_buffer_) {
+    if (frame.pts_ == index_[current_frame_].pts) {
+      if (copy_to_output) {
+        copyD2D(data, frame.frame_.data(), FrameSize());
+      }
+      frame.pts_ = -1;
+
+      ++current_frame_;
+      return true;
+    }
+  }
+
+  current_copy_to_output_ = copy_to_output;
+  current_frame_output_ = data;
+
+  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+    if (av_state_->packet_->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+
+    // Store pts from current packet to indicate,
+    // that this frame is in the decoder
+    piped_pts_.push(av_state_->packet_->pts);
+
+    // Add header needed for NVDECODE to the packet
+    if (filtered_packet_->data) {
+      av_packet_unref(filtered_packet_);
+    }
+    DALI_ENFORCE(av_bsf_send_packet(bsfc_, av_state_->packet_) >= 0);
+    DALI_ENFORCE(av_bsf_receive_packet(bsfc_, filtered_packet_) >= 0);
+
+    // Prepare nv packet
+    CUVIDSOURCEDATAPACKET *packet = &nvdecode_state_->packet;
+    memset(packet, 0, sizeof(CUVIDSOURCEDATAPACKET));
+    packet->payload = filtered_packet_->data;
+    packet->payload_size = filtered_packet_->size;
+    packet->flags = CUVID_PKT_TIMESTAMP;
+    packet->timestamp = filtered_packet_->pts;
+
+    // Send packet to the nv deocder
+    frame_returned_ = false;
+    CUDA_CALL(cuvidParseVideoData(nvdecode_state_->parser_, packet));
+
+    if (frame_returned_) {
+      ++current_frame_;
+      return true;
+    }
+  }
+
+  if (!last_frame_read_) {
```

Comment:
I don't fully understand that. `last_frame_read_` is always false. When do you set it to `true`?

Reply:
Leftover from before. Removed it, since now the same information is passed by `next_frame_idx`.